File: frontend/src/lib/stores/assistant.ts
import { get, writable } from 'svelte/store';
import { apiFetch } from '$lib/api/client';
import { language } from '$lib/i18n';
import type { AssistantEditPlanRequest, AssistantEditPlanResponse, AssistantGalleryBatchJobStatus, AssistantGalleryBatchRequest, AssistantGalleryImageResponse, AssistantGalleryMetadataResponse, AssistantImagePromptOptimizeResponse, AssistantImagePromptResponse, AssistantJobDiagnoseRequest, AssistantJobDiagnoseResponse, AssistantPromptCheckRequest, AssistantPromptCheckResponse, AssistantPromptRewriteRequest, AssistantPromptRewriteResponse, AssistantPromptVariantsRequest, AssistantPromptVariantsResponse, AssistantRecommendParamsRequest, AssistantRecommendParamsResponse } from '$lib/api/types/assistant';

export type AssistantState = {
  promptLoading: boolean;
  paramsLoading: boolean;
  diagnoseLoadingJobId: string | null;
  editPlanLoading: boolean;
  galleryLoadingImageId: string | null;
  promptLoadingCount: number;
  paramsLoadingCount: number;
  editPlanLoadingCount: number;
  galleryLoadingImageIds: string[];
  batchJob: AssistantGalleryBatchJobStatus | null;
};

const initialAssistantState: AssistantState = {
  promptLoading: false,
  paramsLoading: false,
  diagnoseLoadingJobId: null,
  editPlanLoading: false,
  galleryLoadingImageId: null,
  promptLoadingCount: 0,
  paramsLoadingCount: 0,
  editPlanLoadingCount: 0,
  galleryLoadingImageIds: [],
  batchJob: null
};

export function isAbortError(error: unknown): boolean {
  return error instanceof Error && error.name === 'AbortError';
}

function createAssistantStore() {
  const { subscribe, update } = writable<AssistantState>(initialAssistantState);
  const operationControllers = new Map<string, AbortController>();

  function beginCounter(key: 'prompt' | 'params' | 'editPlan') {
    update((state) => {
      if (key === 'prompt') {
        const count = state.promptLoadingCount + 1;
        return { ...state, promptLoadingCount: count, promptLoading: count > 0 };
      }
      if (key === 'params') {
        const count = state.paramsLoadingCount + 1;
        return { ...state, paramsLoadingCount: count, paramsLoading: count > 0 };
      }
      const count = state.editPlanLoadingCount + 1;
      return { ...state, editPlanLoadingCount: count, editPlanLoading: count > 0 };
    });
  }

  function endCounter(key: 'prompt' | 'params' | 'editPlan') {
    update((state) => {
      if (key === 'prompt') {
        const count = Math.max(0, state.promptLoadingCount - 1);
        return { ...state, promptLoadingCount: count, promptLoading: count > 0 };
      }
      if (key === 'params') {
        const count = Math.max(0, state.paramsLoadingCount - 1);
        return { ...state, paramsLoadingCount: count, paramsLoading: count > 0 };
      }
      const count = Math.max(0, state.editPlanLoadingCount - 1);
      return { ...state, editPlanLoadingCount: count, editPlanLoading: count > 0 };
    });
  }

  function beginGalleryLoading(imageId: string) {
    update((state) => {
      const ids = [...state.galleryLoadingImageIds, imageId];
      return { ...state, galleryLoadingImageIds: ids, galleryLoadingImageId: ids.at(-1) ?? null };
    });
  }

  function endGalleryLoading(imageId: string) {
    update((state) => {
      const index = state.galleryLoadingImageIds.indexOf(imageId);
      const ids =
        index >= 0
          ? [
              ...state.galleryLoadingImageIds.slice(0, index),
              ...state.galleryLoadingImageIds.slice(index + 1)
            ]
          : state.galleryLoadingImageIds;
      return { ...state, galleryLoadingImageIds: ids, galleryLoadingImageId: ids.at(-1) ?? null };
    });
  }

  function operationSignal(key: string, signal?: AbortSignal): AbortSignal | undefined {
    if (signal) return signal;
    operationControllers.get(key)?.abort();
    const controller = new AbortController();
    operationControllers.set(key, controller);
    return controller.signal;
  }

  function clearOperationSignal(key: string, signal?: AbortSignal) {
    const controller = operationControllers.get(key);
    if (controller && controller.signal === signal) operationControllers.delete(key);
  }

  async function rewritePrompt(body: AssistantPromptRewriteRequest, signal?: AbortSignal) {
    const requestSignal = operationSignal('prompt', signal);
    beginCounter('prompt');
    try {
      return await apiFetch<AssistantPromptRewriteResponse>(
        '/api/assistant/prompt/rewrite',
        {
          method: 'POST',
          signal: requestSignal,
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(body)
        },
        'rewriting prompt with AI Assistant'
      );
    } finally {
      endCounter('prompt');
      clearOperationSignal('prompt', requestSignal);
    }
  }

  async function checkPrompt(body: AssistantPromptCheckRequest, signal?: AbortSignal) {
    const requestSignal = operationSignal('prompt', signal);
    beginCounter('prompt');
    try {
      return await apiFetch<AssistantPromptCheckResponse>(
        '/api/assistant/prompt/check',
        {
          method: 'POST',
          signal: requestSignal,
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(body)
        },
        'checking prompt with AI Assistant'
      );
    } finally {
      endCounter('prompt');
      clearOperationSignal('prompt', requestSignal);
    }
  }

  async function promptVariants(body: AssistantPromptVariantsRequest, signal?: AbortSignal) {
    const requestSignal = operationSignal('prompt', signal);
    beginCounter('prompt');
    try {
      return await apiFetch<AssistantPromptVariantsResponse>(
        '/api/assistant/prompt/variants',
        {
          method: 'POST',
          signal: requestSignal,
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(body)
        },
        'creating prompt variants with AI Assistant'
      );
    } finally {
      endCounter('prompt');
      clearOperationSignal('prompt', requestSignal);
    }
  }

  async function recommendParams(body: AssistantRecommendParamsRequest, signal?: AbortSignal) {
    const requestSignal = operationSignal('params', signal);
    beginCounter('params');
    try {
      return await apiFetch<AssistantRecommendParamsResponse>(
        '/api/assistant/generate/recommend-params',
        {
          method: 'POST',
          signal: requestSignal,
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(body)
        },
        'recommending generation parameters with AI Assistant'
      );
    } finally {
      endCounter('params');
      clearOperationSignal('params', requestSignal);
    }
  }

  async function diagnoseJob(jobId: string, body: AssistantJobDiagnoseRequest = { include_prompt: false }, signal?: AbortSignal) {
    const requestSignal = operationSignal(`diagnose:${jobId}`, signal);
    update((state) => ({ ...state, diagnoseLoadingJobId: jobId }));
    try {
      return await apiFetch<AssistantJobDiagnoseResponse>(
        `/api/assistant/jobs/${encodeURIComponent(jobId)}/diagnose`,
        {
          method: 'POST',
          signal: requestSignal,
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(body)
        },
        'diagnosing job with AI Assistant'
      );
    } finally {
      update((state) => ({ ...state, diagnoseLoadingJobId: null }));
      clearOperationSignal(`diagnose:${jobId}`, requestSignal);
    }
  }

  async function planEdit(body: AssistantEditPlanRequest, signal?: AbortSignal) {
    const requestSignal = operationSignal('editPlan', signal);
    beginCounter('editPlan');
    try {
      return await apiFetch<AssistantEditPlanResponse>(
        '/api/assistant/edit/plan',
        {
          method: 'POST',
          signal: requestSignal,
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(body)
        },
        'planning edit with AI Assistant'
      );
    } finally {
      endCounter('editPlan');
      clearOperationSignal('editPlan', requestSignal);
    }
  }

  async function promptFromImage(file: File, targetLanguage: 'en' | 'zh-CN', signal?: AbortSignal) {
    const requestSignal = operationSignal('imagePrompt', signal);
    beginCounter('prompt');
    const formData = new FormData();
    formData.append('image', file, file.name);
    formData.append('target_language', targetLanguage);
    try {
      return await apiFetch<AssistantImagePromptResponse>(
        '/api/assistant/image/prompt',
        {
          method: 'POST',
          signal: requestSignal,
          body: formData
        },
        'reverse prompting uploaded image with AI Assistant'
      );
    } finally {
      endCounter('prompt');
      clearOperationSignal('imagePrompt', requestSignal);
    }
  }

  async function optimizeImagePrompt(file: File, prompt: string, targetLanguage: 'en' | 'zh-CN', signal?: AbortSignal) {
    const requestSignal = operationSignal('imagePromptOptimize', signal);
    beginCounter('prompt');
    const formData = new FormData();
    formData.append('image', file, file.name);
    formData.append('prompt', prompt);
    formData.append('target_language', targetLanguage);
    try {
      return await apiFetch<AssistantImagePromptOptimizeResponse>(
        '/api/assistant/image/prompt/optimize',
        {
          method: 'POST',
          signal: requestSignal,
          body: formData
        },
        'generating and comparing a reverse-prompt preview'
      );
    } finally {
      endCounter('prompt');
      clearOperationSignal('imagePromptOptimize', requestSignal);
    }
  }

  async function describeGalleryImage(imageId: string, signal?: AbortSignal) {
    const requestSignal = operationSignal(`gallery:describe:${imageId}`, signal);
    beginGalleryLoading(imageId);
    try {
      return await apiFetch<AssistantGalleryImageResponse>(
        `/api/assistant/gallery/${encodeURIComponent(imageId)}/describe`,
        { method: 'POST', signal: requestSignal },
        'describing gallery image with AI Assistant'
      );
    } finally {
      endGalleryLoading(imageId);
      clearOperationSignal(`gallery:describe:${imageId}`, requestSignal);
    }
  }

  async function promptFromGalleryImage(imageId: string, signal?: AbortSignal) {
    const requestSignal = operationSignal(`gallery:prompt:${imageId}`, signal);
    beginGalleryLoading(imageId);
    try {
      return await apiFetch<AssistantGalleryImageResponse>(
        `/api/assistant/gallery/${encodeURIComponent(imageId)}/prompt?target_language=${encodeURIComponent(get(language))}`,
        { method: 'POST', signal: requestSignal },
        'reverse prompting gallery image with AI Assistant'
      );
    } finally {
      endGalleryLoading(imageId);
      clearOperationSignal(`gallery:prompt:${imageId}`, requestSignal);
    }
  }

  async function analyzeGalleryImage(imageId: string, signal?: AbortSignal) {
    const requestSignal = operationSignal(`gallery:analyze:${imageId}`, signal);
    beginGalleryLoading(imageId);
    try {
      return await apiFetch<AssistantGalleryImageResponse>(
        `/api/assistant/gallery/${encodeURIComponent(imageId)}/analyze?target_language=${encodeURIComponent(get(language))}`,
        { method: 'POST', signal: requestSignal },
        'analyzing gallery image with AI Assistant'
      );
    } finally {
      endGalleryLoading(imageId);
      clearOperationSignal(`gallery:analyze:${imageId}`, requestSignal);
    }
  }

  async function loadGalleryMetadata(imageId: string) {
    return apiFetch<AssistantGalleryMetadataResponse>(
      `/api/assistant/gallery/${encodeURIComponent(imageId)}/metadata`,
      {},
      'loading AI gallery metadata'
    );
  }

  async function batchAnalyzeGallery(body: AssistantGalleryBatchRequest) {
    const job = await apiFetch<AssistantGalleryBatchJobStatus>(
      '/api/assistant/gallery/batch/analyze',
      {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body)
      },
      'starting gallery AI analysis'
    );
    update((state) => ({ ...state, batchJob: job }));
    return job;
  }

  function reset() {
    for (const controller of operationControllers.values()) controller.abort();
    operationControllers.clear();
    update(() => ({ ...initialAssistantState }));
  }

  return {
    subscribe,
    rewritePrompt,
    checkPrompt,
    promptVariants,
    recommendParams,
    diagnoseJob,
    planEdit,
    promptFromImage,
    optimizeImagePrompt,
    describeGalleryImage,
    promptFromGalleryImage,
    analyzeGalleryImage,
    loadGalleryMetadata,
    batchAnalyzeGallery,
    reset
  };
}

export const assistantStore = createAssistantStore();

File: backend/app/api/routers/assistant.py
"""Assistant HTTP request and response mapping."""

from typing import Literal

from fastapi import APIRouter, Body, File, Form, Request, UploadFile

from ..responses import streaming_response
from ..uploads import read_image_upload, resolve_upload_content_type
from ...core import security as auth
from ...core.image_models import MAX_PROMPT_CHARS
from ...schemas.settings import AIAssistantSettingsRequest
from ...schemas.assistant import AssistantImagePromptOptimizeResponse, AssistantImagePromptResponse

from ...schemas.assistant import (
    AssistantEditPlanResponse,
    AssistantGalleryBatchJobStatus,
    AssistantGalleryImageResponse,
    AssistantGalleryMetadataResponse,
    AssistantHealthResponse,
    AssistantImagePromptResponse,
    AssistantImagePromptOptimizeResponse,
    AssistantJobDiagnoseResponse,
    AssistantPromptCheckResponse,
    AssistantPromptRewriteResponse,
    AssistantPromptVariantsResponse,
    AssistantRecommendParamsResponse,
)
from ...services import assistant_batch, assistant_text, assistant_vision

router = APIRouter()

router.add_api_route("/api/assistant/prompt/rewrite", assistant_text.rewrite_prompt, methods=["POST"], response_model=AssistantPromptRewriteResponse)
router.add_api_route("/api/assistant/prompt/check", assistant_text.check_prompt, methods=["POST"], response_model=AssistantPromptCheckResponse)
router.add_api_route("/api/assistant/prompt/variants", assistant_text.prompt_variants, methods=["POST"], response_model=AssistantPromptVariantsResponse)
router.add_api_route("/api/assistant/generate/recommend-params", assistant_text.recommend_generate_params, methods=["POST"], response_model=AssistantRecommendParamsResponse)
router.add_api_route("/api/assistant/jobs/{job_id}/diagnose", assistant_text.diagnose_job, methods=["POST"], response_model=AssistantJobDiagnoseResponse)
router.add_api_route("/api/assistant/edit/plan", assistant_text.plan_edit, methods=["POST"], response_model=AssistantEditPlanResponse)
router.add_api_route("/api/assistant/gallery/{image_id}/metadata", assistant_vision.get_gallery_metadata, methods=["GET"], response_model=AssistantGalleryMetadataResponse)
router.add_api_route("/api/assistant/gallery/batch/analyze", assistant_batch.batch_analyze_gallery, methods=["POST"], response_model=AssistantGalleryBatchJobStatus, status_code=202)
router.add_api_route("/api/assistant/gallery/batch/analyze/{job_id}", assistant_batch.get_batch_analyze_job, methods=["GET"], response_model=AssistantGalleryBatchJobStatus)
router.add_api_route("/api/assistant/gallery/{image_id}/describe", assistant_vision.describe_gallery_image, methods=["POST"], response_model=AssistantGalleryImageResponse)
router.add_api_route("/api/assistant/gallery/{image_id}/prompt", assistant_vision.prompt_gallery_image, methods=["POST"], response_model=AssistantGalleryImageResponse)
router.add_api_route("/api/assistant/gallery/{image_id}/analyze", assistant_vision.analyze_gallery_image, methods=["POST"], response_model=AssistantGalleryImageResponse)


@router.post("/api/assistant/health", response_model=AssistantHealthResponse)
async def assistant_health(req: AIAssistantSettingsRequest | None = Body(default=None)):
    return await assistant_text.assistant_health(req)


@router.post("/api/assistant/image/prompt", response_model=AssistantImagePromptResponse)
async def assistant_image_prompt(
    image: UploadFile = File(...),
    target_language: Literal["en", "zh-CN"] = Form("en"),
):
    image_bytes = await read_image_upload(image)
    return await assistant_vision.prompt_from_uploaded_image(
        image_bytes=image_bytes,
        filename=image.filename or "image",
        content_type=resolve_upload_content_type(image),
        target_language=target_language,
    )


@router.post(
    "/api/assistant/image/prompt/optimize",
    response_model=AssistantImagePromptOptimizeResponse,
)
async def assistant_optimize_image_prompt(
    image: UploadFile = File(...),
    prompt: str = Form(..., min_length=1, max_length=MAX_PROMPT_CHARS),
    target_language: Literal["en", "zh-CN"] = Form("en"),
):
    image_bytes = await read_image_upload(image)
    return await assistant_vision.optimize_uploaded_image_prompt(
        image_bytes=image_bytes,
        filename=image.filename or "image",
        content_type=resolve_upload_content_type(image),
        prompt=prompt,
        target_language=target_language,
    )


@router.get("/api/assistant/gallery/batch/analyze/{job_id}/events")
async def stream_assistant_batch_analyze_job(job_id: str, request: Request):
    body = await assistant_batch.stream_batch_analyze_job(
        job_id,
        client_ip=auth.get_client_ip(request),
        is_disconnected=request.is_disconnected,
    )
    return streaming_response(body)
